apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/RpcRetryConsumer.cs

```diff
@@ -0,0 +1,214 @@
+            }
+        }
+
+        /// <inheritdoc />
+        public Task ReceiveAsync(IEnumerable<T> items, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            GaxPreconditions.CheckNotNull(items, nameof(items));
+            try
+            {
+                return _consumer.ReceiveAsync(items, cancellationToken);
+            }
+            catch (RpcException e)
+            {
+                HandleException(items, e);
+                return CommonUtils.CompletedTask;
+            }
+        }
+
+        /// <inheritdoc />
+        public void Dispose() => _timer?.Dispose();
+
+        /// <summary>
+        /// Handles a <see cref="RpcException"/> that occurred during a <see cref="Receive(IEnumerable{T})"/> call.
+        /// </summary>
+        private void HandleException(IEnumerable<T> items, RpcException e)
+        {
+            // If we are not retrying failed calls either propagate or swallow the exception.
+            if (_options.RetryType == RetryType.None)
+            {
+                switch (_options.ExceptionHandling)
+                {
+                    case ExceptionHandling.Propagate:
+                        throw e;
+                    case ExceptionHandling.Swallow:
+                        Debug.WriteLine($"Swallowed exception: {e}");
+                        return;
+                    default:
+                        throw new InvalidOperationException(
+                            $"Invalid ExceptionHandling option: {_options.ExceptionHandling}");
+                }
+            }
+            
+            // Get the size of items to be buffered and retried.
+            int size = items.Sum(item => _sizer(item));
+
+            // If the new items would be greater than the size of the buffer drop them or throw.
+            if (size + _size > _options.BufferSizeBytes) 
+            {
+                switch (_options.BufferOverflow)
+                {
+                    case BufferOverflow.IgnoreNewEntries:
+                        Debug.WriteLine($"Ignored {items.Count()} entries.");
+                        return;
+                    case BufferOverflow.ThrowException:
+                        throw new InvalidOperationException("Buffer overflow");
+                    default:
+                        throw new InvalidOperationException(
+                            $"Invalid BufferOverflow option: {_options.BufferOverflow}");
+                }
+            }
+
+            // Lock the semaphore and add the new items to the buffer.
+            _semaphore.Wait();
+            _size += size;
+            _buffer.AddRange(items);
+            _semaphore.Release();
+        }
+
+        /// <summary>
+        /// Try to flush the buffer if it has any entries.
+        /// </summary>
+        private void FlushBuffer()
+        {
+            // If the buffer is empty do nothing.
+            if (_buffer.Count() == 0)
```

Comment:
This should be in the lock as well or you should do double-checked locking. Two threads could each check this and pass it before one of them resets the buffer. Then the second thread will flush an empty buffer.